### ISSUE TITLE
Nested map blocks are not displayed in plan diff

### ISSUE THREAD

**Issue by ebekker** [hashicorp/terraform#20902]
### Terraform Version

```
Terraform v0.12.0-beta1
```

### Terraform Configuration Files

Sample _baseline_ TF config using a custom provider (`winlo`):

```hcl
terraform {
    required_version = ">= 0.11.13"
}

provider "winlo" {
}

resource "winlo_registry_key" "sample2" {
    ## `entry_for_list` is a nested block of schema type `list`
    entry_for_list {
        type = "string"
    }

    ## `entry_for_map` is a nested block of schema type `map`
    entry_for_map "entry_1" {
        type = "string"
    }

    ## the element type and schema for both of
    ## the above nested blocks is the same
}
```

Sample TF config with added nested block **list** entry:

```hcl
terraform {
    required_version = ">= 0.11.13"
}

provider "winlo" {
}

resource "winlo_registry_key" "sample2" {
    entry_for_list {
        type = "string"
    }

    ## *new* entry here
    entry_for_list { 
        type = "string"
    }

    entry_for_map "entry_1" {
        type = "string"
    }
}
```

This config generates the following *plan* output during an apply (as expected):

```
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # winlo_registry_key.sample2 will be updated in-place
  ~ resource "winlo_registry_key" "sample2" {
        path = "Terrascape\\SampleKey2"
        root = "HKCU"

        entry_for_list {
            type = "string"
        }
      + entry_for_list {
          + type = "string"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

Sample TF config with added nested block **map** entry:

```hcl
terraform {
    required_version = ">= 0.11.13"
}

provider "winlo" {
}

resource "winlo_registry_key" "sample2" {
    entry_for_list {
        type = "string"
    }

    entry_for_map "entry_1" {
        type = "string"
    }

    ## *new* entry here
    entry_for_map "entry_2" {
        type = "string"
    }
}
```

This config generates the following *plan* output during an apply, *but* is missing the diff of the new nested block map entry:

```
Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # winlo_registry_key.sample2 will be updated in-place
  ~ resource "winlo_registry_key" "sample2" {
        path = "Terrascape\\SampleKey2"
        root = "HKCU"

        entry_for_list {
            type = "string"
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

### Debug Output

Debug output for the two sample `apply` runs from above:
* [adding-nested-block-list-item](https://gist.github.com/ebekker/38b6b7a93d84aa9cc5b1ae9e432dd90c#file-terraform0-12-beta1-trace-adding-nested-block-list-item)
* [adding-nested-block-map-entry](https://gist.github.com/ebekker/38b6b7a93d84aa9cc5b1ae9e432dd90c#file-terraform0-12-beta1-trace-adding-nested-block-map-entry)

### Expected Behavior

Expected to see the nested block map entries to be represented in the plan diffs.

### Actual Behavior

Only the nested block list entries are displayed in the plans, no nested block map entries.

### Steps to Reproduce

Create a provider plugin with a schema that defines a nested block block of Map type.  Create a sample TF configuration using that provider and try to `plan` or `apply`.

### Additional Context

I am invoking the TF comman from a wrapper PWSH script.  The Provider is also ***not*** implemented in Golang, but this doesn't seem to be a contributing factor.

### References

Possibly related:
* https://github.com/hashicorp/terraform/issues/20076

**Comment by apparentlymart** [comment]
Hi @ebekker! Thanks for reporting this.

It looks like you've found the same issue I was working on in #20630, which was eventually fixed as part of #20631. That was merged after beta1 was already out, so it'll be reflected in the forthcoming beta2, or if you like you could check with a build from latest `master` to see.

I'm a little unsure though how exactly you built a provider using this feature with the current Terraform SDK. The capability for blocks to be backed by maps was introduced in Terraform v0.12 but the official Terraform SDK doesn't support it yet, and cannot support it until v0.11 compatibility is ended. Are you implementing this provider in an unusual way (e.g. without an SDK or with a custom one), or are you perhaps using an older vendored copy of the SDK from before #20116 was merged?

(When I was working on #20630 I was using a contrived test provider that bypasses the official SDK and just implements the wire protocol directly, which is how I was able to get the screenshot illustrating it working correctly.)

**Comment by ebekker** [comment]
Ah, that's great, thanks for confirming.

So the secret is...my provider is not written using the SDK, it's not even written in Golang.

I'm writing a provider in C#/.NET Core as part of [this little experiement](https://github.com/ebekker/terrascape).

**Comment by ebekker** [comment]
@apparentlymart -- if I wanted to get the latest build from `master` to test this out, where do I find pre-release builds?

**Comment by apparentlymart** [comment]
There is not currently a pre-release build containing that fix. You'd need to build from `master` yourself, or wait for the forthcoming `beta2` release (which is currently holding for more providers having releases that support protocol version 5).

Thanks for confirming that you're not using the official SDK. On re-read of your message I see that you did mention that at the end but I missed that detail on my first read-through. I'm glad to see some experimentation with SDKs in other languages; that was one of the goals of switching over to `grpc`, after all. I'm sure you saw this already, but just be clear for anyone else who is following along here: protocol version 5 is not frozen until v0.12.0 final is released, so the `.proto` file is subject to change in incompatible ways until that time. After that, there will be only backward-compatible changes until protocol version 6.

You're blazing ahead of us having documented fully the new protocol yet, so please note that moving forward we are expecting plugin servers to be able to support multiple protocol major versions at once using a negotiation protocol [based on the `PLUGIN_PROTOCOL_VERSIONS` environment variable](https://github.com/hashicorp/go-plugin/blob/52e1c4730856c1438ced7597c9b5c585a7bd06a2/server.go#L99). Since your implementation currently supports only protocol 5 that's not a big deal for now, but I mention this just in case you'd like to incorporate the possibility of this into your library API. The main implication of it is that the server implementation would accept potentially multiple supported application protocol versions and choose the newest supported one that's listed in `PLUGIN_PROTOCOL_VERSIONS`, presumably then selecting a different GRPC service stub based on the negotiated version.

**Comment by ebekker** [comment]
Thanks for the heads up!

I'll keep an eye out for changes to the `.proto` file and make the necessary updates to keep pace.

I'll keep the proto version negotiation in mind as I test out future TF releases.  In general, I've made a decision to only support the newer editions of the various components I'm dependent on (both TF binary and the .NET Core 3.0 runtime which is also in [pre-release](https://dotnet.microsoft.com/download/dotnet-core/3.0) as of now).

As I mentioned, for now this is mostly an experiment, but if it _has legs_ I'll pby need to revisit some of the choices I've made, but by then TF 0.12 final will pby be released :-)
